## Lab 03.01 - Introduction to Pandas and Data Operations
In this lab, you'll learn how to use the pandas library to analyze real-world sales data. By the end of this lab, you'll be comfortable with the fundamental operations in pandas.

### Part 0 - Intro to Pandas
Before we dive into using pandas, let's understand what it is and why it's important in data science.

To answer the questions, edit the markdown cell and put your answer below the question. 

**Make sure to save the markdown cell, by pressing the ✓ (check) icon in the top right after answering the questions**

Research and critically think about the following questions:

##### Question 00
What is Pandas, and what are its primary functions in data visualization for data science?
- **Answer:**
Pandas is a Python library made to work with data. It allows users to create data structures and perform mathematical operations on numeric data.

##### Question 01
What does "data manipulation" mean in the context of pandas and data science field?
- **Answer:**
Data manipulation means the ways data can be altered or worked with in the context of the data science field and Pandas. 
##### Question 02
What kind of files can pandas read? List at least 3:
- **Answer:**
Pandas can read multiple different file types such as csv, json, sql, etc.

#### 0.0 - Hands-On Exploration
Let's start by importing pandas and exploring its basic functionality. Type these commands and write down what you observe:

First, let's install the matplotlib library, in your terminal run the following command:

```bash
pip3 install pandas
```

Once installed, let's start by importing pandas and exploring its basic functionality. 

In [1]:
import pandas as pd

Now lets run create a list, and see what happens when we use it with pandas.

In [2]:
# Create a simple list of numbers
numbers = [1, 2, 3, 4, 5]

# Convert it to a pandas Series
series = pd.Series(numbers)

# Print the result
print(series)

0    1
1    2
2    3
3    4
4    5
dtype: int64


##### Question 03
What did you notice about the output format?
- **Answer:**
Pandas prints data in columns as the index numbers are in their own column and the list items are in a second column.

##### Question 04
How is this different from a regular Python list? What are the numbers on the left side?
- **Answer:**
Regular Python lists are printed as they are written in the code. Lists are printed with the brackets (or other container) the list has and commas that separate the items.

### Part 1 - Reading and Exploring Data

#### 01.00 - Loading Data
Let's start by loading our sales data that you downloaded alongside this lab. In pandas, we can read various file formats, but CSV (Comma-Separated Values) is one of the most common.

In [3]:
sales_df = pd.read_csv('sales_data.csv')

When using pandas to load data, in ingests it into what is known as a "dataframe", this allows us to do advanced manipulation.

#### 01.01 - Viewing Data Samples
Let's explore the fundamental methods for understanding our data structure and content. We'll look at each method individually and understand what it tells us about our data.

In [6]:
print(sales_df.head(3))

         Date   Product         Category    Price  Units    Total  Discount  \
0  2023-01-01  Notebook  Office Supplies    18.33      2    35.56      0.03   
1  2023-01-01    Laptop      Electronics  1943.35      4  7773.40      0.00   
2  2023-01-01    Tablet      Electronics   436.05      2   872.10      0.00   

  Region     Sales_Rep Payment_Method Customer_Segment  
0  South  Sarah Wilson     Debit Card   Small Business  
1  North    John Smith     Debit Card   Small Business  
2   East   David Brown  Bank Transfer       Government  


In [5]:
print(sales_df.tail())

            Date       Product         Category    Price  Units    Total  \
3766  2023-12-31    Smartwatch      Electronics  1730.72      2  3461.44   
3767  2023-12-31          Desk        Furniture   462.57      1   462.57   
3768  2023-12-31        Laptop      Electronics   560.79      2  1121.58   
3769  2023-12-31  Office Chair        Furniture   116.04      2   232.08   
3770  2023-12-31      Notebook  Office Supplies    18.15      4    72.60   

      Discount Region        Sales_Rep Payment_Method Customer_Segment  
3766       0.0  South  Michael Johnson     Debit Card       Individual  
3767       0.0  North       John Smith     Debit Card       Government  
3768       0.0   East      David Brown     Debit Card   Small Business  
3769       0.0  North       Emma Davis    Credit Card        Corporate  
3770       0.0  North       Emma Davis         PayPal       Individual  


##### Question 04
What is the difference between the `tail()` and `head()` commands? What would be use of either command be?
- **Answer:**
The `tail()` function prints the last five data entries and the `head()` function prints the first five entries.

##### Question 05
What happens when you put a number in the `head()` function? What changed?
- **Answer:**
Putting a number in the `head()` function prints that amount of data entries.

#### 01.02 - Understanding DataFrame Structure
Let's examine the basic properties of our DataFrame:

In [7]:
print(sales_df.columns)

Index(['Date', 'Product', 'Category', 'Price', 'Units', 'Total', 'Discount',
       'Region', 'Sales_Rep', 'Payment_Method', 'Customer_Segment'],
      dtype='object')


In [8]:
print(sales_df.dtypes)

Date                 object
Product              object
Category             object
Price               float64
Units                 int64
Total               float64
Discount            float64
Region               object
Sales_Rep            object
Payment_Method       object
Customer_Segment     object
dtype: object


##### Question 06
Compare and contrast the outputs of the `columns` and `dtypes` properties. What is similar what is different?
- **Answer:**
The columns property prints the labels of the dataset in Python list format. The dtypes property prints the labels and what type of data falls under it in columns.
##### Question 07
What data type is the 'Date' column? Is this what you expected? 
- **Answer:**
I am surprised that the data type is considered an object and not an integer but it makes sense as if it were an integer the dates would be wrong as Python would subract the numbers.
##### Question 08
Why might pandas choose different data types for different columns?
- **Answer:**
Pandas might choose different data types between columns to maintain order and ensure the data is formatted properly. 

#### 01.03 - Data Information Summary
The `info()` method provides a concise summary of our DataFrame:

In [9]:
print(sales_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3771 entries, 0 to 3770
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date              3771 non-null   object 
 1   Product           3771 non-null   object 
 2   Category          3771 non-null   object 
 3   Price             3771 non-null   float64
 4   Units             3771 non-null   int64  
 5   Total             3771 non-null   float64
 6   Discount          3771 non-null   float64
 7   Region            3771 non-null   object 
 8   Sales_Rep         3693 non-null   object 
 9   Payment_Method    3739 non-null   object 
 10  Customer_Segment  3771 non-null   object 
dtypes: float64(3), int64(1), object(7)
memory usage: 324.2+ KB
None


##### Question 09
What information does `info()` tell us, how could this information be useful?
- **Answer:**
The `info()` function displays the column names, index numbers, data types, non-null count, list of all data types in data, and how much memory is being used.

##### Question 10
How much memory is our DataFrame using? 
- **Answer:**
Our DataFrame is using 324.2+ KB of data.

##### Question 11
What does "null" mean? What impact could having "null" values in a dataset?
- **Answer:**
"Null" means nothing and having a nothing value can throw off data manipulations as there will be nothing to work with.

#### 01.04 - Numerical Summaries
The `describe()` method provides statistical summaries for numerical columns:

In [10]:
print(sales_df.describe())

             Price        Units         Total     Discount
count  3771.000000  3771.000000   3771.000000  3771.000000
mean    397.093911     2.985150   1159.701854     0.021594
std     514.524434     1.410003   1750.424109     0.041496
min       4.810000     1.000000      4.990000     0.000000
25%      40.165000     2.000000    110.730000     0.000000
50%     134.860000     3.000000    372.870000     0.000000
75%     516.490000     4.000000   1472.525000     0.020000
max    2073.680000     5.000000  10151.100000     0.150000


##### Question 12
What statistics does the `describe()` function give us?
- **Answer:**
The `describe()` function gives us the amount of entries in the table, mean, standard deviation, minimum and maximum values, and the quartiles.
##### Question 13
Which columns did the `describe()` function perform statistics on? Why? (Hint: why didnt we see 'Product' in the output?)
- **Answer:**
The columns that were given were floats and integers as they are numeric values in the dataset and the mathematical functions needed to get these values can be performed.

In [11]:
print(sales_df[['Price', 'Units']].describe())

             Price        Units
count  3771.000000  3771.000000
mean    397.093911     2.985150
std     514.524434     1.410003
min       4.810000     1.000000
25%      40.165000     2.000000
50%     134.860000     3.000000
75%     516.490000     4.000000
max    2073.680000     5.000000


##### Question 14
What is different in the codeblock above compared to the previous one. How did this affect the output?
- **Answer:**
This code block specifies the columns they want described and as a result the function gives them the numeric information of those columns.

In [12]:
print(sales_df['Category'].unique())

['Office Supplies' 'Electronics' 'Accessories' 'Furniture']


##### Question 15
How many unique product categories are in our data set?
- **Answer:**
There are four unique product categories in our data set.

In [13]:
print("Output 1:")
print(sales_df['Category'].value_counts())

print('')
print('')

print("Output 2:")
print(sales_df['Category'].value_counts(normalize=True))

Output 1:
Category
Office Supplies    953
Accessories        942
Furniture          941
Electronics        935
Name: count, dtype: int64


Output 2:
Category
Office Supplies    0.252718
Accessories        0.249801
Furniture          0.249536
Electronics        0.247945
Name: proportion, dtype: float64


##### Question 16
What is the difference between the two outputs?
- **Answer:**
The first output gives use the product categories, the amount of times it shows up in the data set, and its data type. The second output also prints the product categories but displays their proportions.

##### Question 17
What is our most common product category?
- **Answer:**
The most common product category is office supplies as it occurs 953 times in the dataset.

##### Excercise 00
Using what you've learned, create a code cell direclty below this one to answer these questions:

- What is the date range of our sales data? (Hint: try `min()` and `max()` functions)
- How many unique products do we sell?
- ~~What is the average price by category?~~
- What is our most common payment method?

In [15]:
# What is our most common payment method? - Bank Transfer is in the dataset 976 times.

print(sales_df['Payment_Method'].value_counts())

Payment_Method
Bank Transfer    976
Debit Card       932
Credit Card      923
PayPal           908
Name: count, dtype: int64


### Part 2 - Data Selection and Filtering

#### 02.01 - Column Selection
There are multiple ways to select columns in pandas. Let's explore them:

In [16]:
# Select single column
prices = sales_df['Price']

# Select multiple columns
product_info = sales_df[['Product', 'Price', 'Units']]

##### Question 17
Compare and contrast how we select multiple columns vs a single column.
- **Answer:**
The single and multiple column selection methods both assign the output to a variable and use `sales_df[]` to select columns. The difference is that for single column selection, the value is the desired column in quotes, but for multiple selections the values are formatted like a list within the `sales_df[]` function.

##### Excercise 01
Using what you've learned, create a code cell direclty below this one to select the 'Region' and 'Category' columns and store them in a variable called 'location_data'.

#### 02.01 - Filtering Data
Let's learn how to filter our data based on conditions:

In [17]:
expensive_items = sales_df[sales_df['Price'] > 500]

north_sales = sales_df[sales_df['Region'] == 'North']

print("Expensive items:")
print(expensive_items)
print("\nNorth region sales:")
print(north_sales)

Expensive items:
            Date     Product     Category    Price  Units    Total  Discount  \
1     2023-01-01      Laptop  Electronics  1943.35      4  7773.40      0.00   
9     2023-01-01  Smartwatch  Electronics   893.71      3  2681.13      0.00   
10    2023-01-01      Laptop  Electronics   802.12      3  2093.53      0.13   
13    2023-01-02  Smartphone  Electronics   687.98      3  2063.94      0.00   
16    2023-01-02  Smartwatch  Electronics   507.91      4  2031.64      0.00   
...          ...         ...          ...      ...    ...      ...       ...   
3760  2023-12-31  Smartphone  Electronics  1766.20      4  7064.80      0.00   
3763  2023-12-31      Tablet  Electronics  1424.20      5  7121.00      0.00   
3764  2023-12-31  Smartphone  Electronics   857.34      5  4286.70      0.00   
3766  2023-12-31  Smartwatch  Electronics  1730.72      2  3461.44      0.00   
3768  2023-12-31      Laptop  Electronics   560.79      2  1121.58      0.00   

     Region          S

##### Question 18
What is `sales_df['Price'] > 500` accomplishing in our code? How might we use this in data science workflows?
- **Answer:**
`sales_df['Price'] > 500` only gives us the data entries whose prices are above 500. This is useful in data science as it allows us to filter out data entries and focus or work with the ones we want.

##### Question 19
What logic operator would we use to combine multiple conditonals when filtering columns?
- **Answer:**
We can use the `and`/`or` logic operators to combine multiple conditionals when filtering through data.

##### Excercise 02
Using what you've learned, create a code cell direclty below this one to create the following filters:
- The number of units sold was greater than 5
- The product category is 'Electronics'

In [20]:
elect_units = sales_df['Electronics' and sales_df['Units'] > 5]
print(elect_units)

Empty DataFrame
Columns: [Date, Product, Category, Price, Units, Total, Discount, Region, Sales_Rep, Payment_Method, Customer_Segment]
Index: []


#### Part 3 - Final Summary
Use your experience completing this lab, and the code block below to answer the following summary questions

##### Question 20
In your own words, explain what pandas is and why it's useful for data analysis:
- **Answer:**

##### Question 21
How many unique products are in the dataset?
- **Answer:**

##### Question 22
What was the most popular product category in the `North` region? 
- **Answer:**